### DISCUSSION TITLE
Graceful shutdown - how to use it?

### DISCUSSION THREAD

**Discussion by aerott** [vercel/next.js#38735] [Help] [answered]
Hey, community!

With NextJs v12.2.0 we received long awaited option by me - "graceful shudwons" ( https://github.com/vercel/next.js/pull/36909/ ).

In accordance with instruction, I craeted file "pages/_document.js" with following code:

```
if (process.env.NEXT_MANUAL_SIG_HANDLE) {

  // this should be added in your custom _document
  process.on('SIGTERM', () => {
    
	function myfunc() {
      console.log('boom SIGTERM! after 5s');
    }
    console.log('Received SIGTERM: ', 'cleaning up');
    const myTimeout = setTimeout(myfunc, 5000);
    
    process.exit(0);
  });
  process.on('SIGINT', () => {
	function myfunc() {
      console.log('boom SIGINT! after 5s');
    }
    console.log('Received SIGINT: ', 'cleaning up');
    const myTimeout = setTimeout(myfunc, 5000);
    
    process.exit(0);
  });
}
```

The problem is - my setTimeout function is never executed :(

My console output:
`console.log('Received SIGINT: ', 'cleaning up');`

There is lack of this line : `console.log('boom SIGINT! after 5s');`

What am I doing wrong here?
@ijjk, maybe you can help me 😓

**Comment by icyJoseph** [comment, marked as answer]
Hi,

AFAIK, the moment you call `process.exit` that's it. 

From the [Node.js docs](https://nodejs.org/api/process.html#processexitcode):

> Calling process.exit() will force the process to exit as quickly as possible even if there are still asynchronous operations pending that have not yet completed fully, including I/O operations to process.stdout and process.stderr.

What you can do is move the call to process.exit inside your timeout. 

All in all, this is a sync and async code behaviour that often trips unfamiliar developers. Personally, I had to read, and double check every time I made a statement on this answer.

## Edit, code snippet

```js
const terminator = (signal: string) => {
  console.log(`Received ${signal}: ", "cleaning up`);
  setTimeout(
    (...args) => {
      console.log(...args);
      process.exit(0);
    },
    400,
    "Message"
  );
  // un-commenting this line will cause everything to just stop and kill the pending timeout
  // process.exit(0);
};

[
  "SIGHUP",
  "SIGINT",
  "SIGQUIT",
  "SIGILL",
  "SIGTRAP",
  "SIGABRT",
  "SIGBUS",
  "SIGFPE",
  "SIGUSR1",
  "SIGSEGV",
  "SIGUSR2",
  "SIGTERM",
].forEach(function (sig) {
  process.on(sig, function () {
    terminator(sig);
    console.log("signal: " + sig);
  });
});
```

I post it like this, without any next.js context, so that it is clear that this is Node.js related, rather than inherent to Next.js.

**Reply by aerott** [reply]
Holy Sh*t! Rookie mistake 😅 Many thanks @icyJoseph
